pkg/controller/egress/controller.go

```diff
@@ -0,0 +1,230 @@
+	for _, pod := range pods {
+		// Ignore Pod if it's not scheduled or not running.
+		// TODO: If a Pod is scheduled but not running, it can be included in the EgressGroup so that the agent can
+		// install its SNAT rule right after the Pod's CNI request is processed, which just requires a notification from
+		// CNIServer to the agent's EgressController. However the current notification mechanism (the entityUpdate
+		// channel) allows only single consumer. Once it allows multiple consumers, we can change the condition to
+		// include scheduled Pods that have no IPs.
+		if pod.Spec.NodeName == "" || len(pod.Status.PodIPs) == 0 {
+			continue
+		}
+		podNum++
+		podSet := memberSetByNode[pod.Spec.NodeName]
+		if podSet == nil {
+			podSet = controlplane.GroupMemberSet{}
+			memberSetByNode[pod.Spec.NodeName] = podSet
+		}
+		groupMember := &controlplane.GroupMember{
+			Pod: &controlplane.PodReference{
+				Name:      pod.Name,
+				Namespace: pod.Namespace,
+			},
+		}
+		podSet.Insert(groupMember)
+		// Update the NodeNames in order to set the SpanMeta for EgressGroup.
+		nodeNames.Insert(pod.Spec.NodeName)
+	}
+	updatedEgressGroup := &antreatypes.EgressGroup{
+		UID:               egressGroup.UID,
+		Name:              egressGroup.Name,
+		GroupMemberByNode: memberSetByNode,
+		SpanMeta:          antreatypes.SpanMeta{NodeNames: nodeNames},
+	}
+	klog.V(2).Infof("Updating existing EgressGroup %s with %d Pods on %d Nodes", key, podNum, nodeNames.Len())
+	c.egressGroupStore.Update(updatedEgressGroup)
+	return nil
+}
+
+func (c *EgressController) enqueueEgressGroup(key string) {
+	klog.V(4).Infof("Adding new key %s to EgressGroup queue", key)
+	c.queue.Add(key)
+}
+
+// addEgress processes Egress ADD events and creates corresponding EgressGroup.
+func (c *EgressController) addEgress(obj interface{}) {
+	egress := obj.(*egressv1alpha1.Egress)
+	klog.Infof("Processing Egress %s ADD event with selector (%s)", egress.Name, egress.Spec.AppliedTo)
+	// Create an EgressGroup object corresponding to this Egress and enqueue task to the workqueue.
+	egressGroup := &antreatypes.EgressGroup{
+		Name: egress.Name,
+		UID:  egress.UID,
+	}
+	c.egressGroupStore.Create(egressGroup)
+	// Register the group to the grouping interface.
+	groupSelector := antreatypes.NewGroupSelector("", egress.Spec.AppliedTo.PodSelector, egress.Spec.AppliedTo.NamespaceSelector, nil)
+	c.groupingInterface.AddGroup(egressGroupType, egress.Name, groupSelector)
+	c.queue.Add(egress.Name)
+}
+
+// updateEgress processes Egress UPDATE events and updates corresponding EgressGroup.
+func (c *EgressController) updateEgress(old, cur interface{}) {
+	oldEgress := old.(*egressv1alpha1.Egress)
+	curEgress := cur.(*egressv1alpha1.Egress)
+	klog.Infof("Processing Egress %s UPDATE event with selector (%s)", curEgress.Name, curEgress.Spec.AppliedTo)
+	// Do nothing if AppliedTo doesn't change.
+	if reflect.DeepEqual(oldEgress.Spec.AppliedTo, curEgress.Spec.AppliedTo) {
```

Comment:
Do you think we can optimize by defining our own comparison funcs for AppliedTo?

Reply:
Sure, will do

Reply:
We can do later.